stix2/sources/__init__.py

```diff
@@ -557,150 +687,198 @@ def add_filter(self, filters):
     def deduplicate(self, stix_obj_list):
-        """deduplicate a list of STIX objects into a unique set
+        """deduplicate a list fo STIX objects to a unique set
 
-        reduces a set of STIX objects to unique set by looking
+        Reduces a set of STIX objects to unique set by looking
         at 'id' and 'modified' fields - as a unique object version
         is determined by the combination of those fields
 
         Args:
             stix_obj_list (list): list of STIX objects (dicts)
 
         Returns:
-            (list): a unique set of the passed STIX object list
-
-
+                (list): unique set of the passed list of STIX objects
         """
+
         unique = []
-        have = False
+        dont_have = False
         for i in stix_obj_list:
+            dont_have = False
             for j in unique:
-                if i['id'] == j['id'] and i['modified'] == j['modified']:
-                    have = True
+                for field in STIX_VERSION_FIELDS:
+                    if not i[field] == j[field]:
+                        dont_have = True
                     break
-            if not have:
+            if dont_have:
                 unique.append(i)
-            have = False
         return unique
+
+
+class STIXCommonPropertyFilters():
+    """
+    """
+    @classmethod
+    def _all(cls, filter_, stix_obj_field):
+        """all filter operations (for filters whose value type can be applied to any operation type)"""
+        if filter_["op"] == '=':
+            return stix_obj_field == filter_["value"]
+        elif filter_["op"] == "!=":
+            return stix_obj_field != filter_["value"]
+        elif filter_["op"] == "in":
+            return stix_obj_field in filter_["value"]
+        elif filter_["op"] == ">":
+            return stix_obj_field > filter_["value"]
+        elif filter_["op"] == "<":
+            return stix_obj_field < filter_["value"]
+        elif filter_["op"] == ">=":
+            return stix_obj_field >= filter_["value"]
+        elif filter_["op"] == "<=":
+            return stix_obj_field <= filter_["value"]
+        else:
+            return -1
+
+    @classmethod
+    def _id(cls, filter_, stix_obj_id):
+        """base filter types"""
+        if filter_["op"] == "=":
+            return stix_obj_id == filter_["value"]
+        elif filter_["op"] == "!=":
+            return stix_obj_id != filter_["value"]
+        else:
+            return -1
+
+    @classmethod
+    def _boolean(filter_, stix_obj_field):
+        if filter_["op"] == "=":
+            return stix_obj_field == filter_["value"]
+        elif filter_["op"] == "!=":
+            return stix_obj_field != filter_["value"]
+        else:
+            return -1
+
+    @classmethod
+    def _string(cls, filter_, stix_obj_field):
+        return cls._all(filter_, stix_obj_field)
+
+    @classmethod
+    def _timestamp(cls, filter_, stix_obj_timestamp):
+        return cls._all(filter_, stix_obj_timestamp)
+
```

Comment:
Type-specific checks...  Feels familiar: the matcher does that, since it has to check property values of particular types from a JSON object.  Similar considerations could come into play with these filters.  E.g. one could decide that if types are incompatible, that's a failed match but not an error, just like the pattern spec.  Might people ever want to filter in ways analogous to how patterns work?  E.g. have IP addresses and subnet tests?  Regex matches on strings?  Something to think about.

Filters are supposed to operate on our SDO python types, right?  Should one be able to determine property types from an SDO instance?  E.g. look up in _properties?  But perhaps _properties should be considered an implementation detail.  Could a method be written into _STIXBase to determine a type given a property name?  Then all you'd need is a type->tester_function mapping.

Although "identifier" is defined as a distinct type (section 2.5), I'm not sure that it merits its own type for filtering.  Wouldn't plain old string tests do the trick?  Is `STIXCommonPropertyFilters._id` necessary?  Would it be so bad to allow all of the string ops on an ID (even if they would probably mostly never be used)?